### PR TITLE
Introduce a function for adding custom tags to system metrics

### PR DESCRIPTION
**Pull request by LakshanKarunathilake**
Current API comprises one endpoint which can be used to add tags to tracing spans. We don't have an endpoint to add a tag to system metrics.

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/25395

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [x] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: observelib/observe/src/main/java/org/ballerinalang/observe/nativeimpl/AddTagToMetrics.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    public static Object addTagToMetrics(Environment env, BString tagKey, BString tagValue) {
+
+        ObserverContext observerContext = ObserveUtils.getObserverContextOfCurrentFrame(env);
+        boolean isCustomTagsAvailable = true;
+        if (observerContext != null) {
+            Map<String, Tag> customTags =
+                    (Map<String, Tag>) observerContext.getProperty(BallerinaMetricsObserver.PROPERTY_CUSTOM_TAGS);
+            if (customTags == null) {
+                customTags = new HashMap<>();
+                isCustomTagsAvailable = false;
```
Comment (nadundesilva) [inline observelib/observe/src/main/java/org/ballerinalang/observe/nativeimpl/AddTagToMetrics.java:50]: U can directly add the property here without maintaining a separate `isCustomTagsAvailable` variable

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/metrics/BallerinaMetricsObserver.java [hunk 2]
```diff
@@ -22,15 +22,23 @@
...
 import java.time.Duration;
+import java.util.HashSet;
+import java.util.Map;
 import java.util.Set;
 
+import static io.ballerina.runtime.observability.ObservabilityConstants.PROPERTY_KEY_HTTP_STATUS_CODE;
+import static io.ballerina.runtime.observability.ObservabilityConstants.STATUS_CODE_GROUP_SUFFIX;
+import static io.ballerina.runtime.observability.ObservabilityConstants.TAG_KEY_HTTP_STATUS_CODE_GROUP;
+
+
```
Comment (nadundesilva) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/metrics/BallerinaMetricsObserver.java:33]: Let's remove this additional line here

File: tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/observability/metrics/MetricsTestCase.java [hunk 3]
```diff
@@ -125,21 +125,41 @@ private Metrics filterByTag(Metrics metrics, String key, String value) {
...
+    }
+
+    /**
+     * Test the metrics generated for a particular function invocation for a given start and end tag set.
+     *
+     * @param allMetrics           All the metrics collected from Metrics Registry
+     * @param invocationPosition   The invocation position of the function invocation
+     * @param invocationCount      The number of times the function should have been called
+     * @param startObservationTags tags at the observation start which should be present in the metrics
+     * @param endObservationTags   tags at the observation end which should be present in the metrics
```
Comment (nadundesilva) [inline tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/observability/metrics/MetricsTestCase.java:146]: We should rename the param and rephrase the description. `endObservationTags` is actually the additional tags added after start here.

File: tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/observability/metrics/MetricsTestCase.java [hunk 4]
```diff
@@ -175,25 +195,28 @@ private void testFunctionCounters(long invocationCount, Metrics functionMetrics,
...
 
     /**
-     * Test the gauge metrics generated for a particular function invocation.
+     * Test the gauge metrics generated for a particular function invocation for given start and end tag set.
      *
      * @param invocationCount The number of times the function should have been called
      * @param functionMetrics All the metrics generated for the function invocation
-     * @param tags Tags that should be present in metrics
+     * @param startTags            Tags that should be present in metrics at observation start
+     * @param endTags            Tags that should be present in metrics at observation end
```
Comment (nadundesilva) [inline tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/observability/metrics/MetricsTestCase.java:203]: Minor alignment issue here in the javadoc.

File: tests/jballerina-integration-test/src/test/resources/observability/metrics_tests/03_addTagToMetrics_function.bal [hunk 5]
```diff
@@ -0,0 +1,29 @@
+// Copyright (c) 2020 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (nadundesilva) [inline tests/jballerina-integration-test/src/test/resources/observability/metrics_tests/03_addTagToMetrics_function.bal:1]: Shall we rename this file to `03_custom_metric_tags.bal` ? (The convention for file names here is "snake_case")

File: tests/jballerina-integration-test/src/test/resources/observability/metrics_tests/03_addTagToMetrics_function.bal [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+// under the License.
+
+import ballerina/testobserve;
+import ballerina/observe;
+
+service testServiceTwo on new testobserve:Listener(10092) {
+    # Resource function for test custom tags
+    resource function testAddTagToMetrics(testobserve:Caller caller) {
+         // Add a custom tag to span, this should not be included in system metrics
+         var res =  observe:addTagToSpan("tracing", "Tracing Value" );
```
Comment (nadundesilva) [inline tests/jballerina-integration-test/src/test/resources/observability/metrics_tests/03_addTagToMetrics_function.bal:24]: Minor alignment and spacing issues here

**Review by nadundesilva — APPROVED** [review]
LGTM

**Review by nadundesilva — DISMISSED** [review]
LGTM

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/metrics/BallerinaMetricsObserver.java [hunk 7]
```diff
@@ -22,15 +22,22 @@
...
+import static io.ballerina.runtime.observability.ObservabilityConstants.TAG_KEY_HTTP_STATUS_CODE_GROUP;
+
 /**
  * Observe the runtime and collect measurements.
  */
 public class BallerinaMetricsObserver implements BallerinaObserver {
 
     private static final String PROPERTY_START_TIME = "_observation_start_time_";
     private static final String PROPERTY_IN_PROGRESS_COUNTER = "_observation_in_progress_counter_";
+    public static final String PROPERTY_CUSTOM_TAGS = "_custom_metric_tags_";
```
Comment (nadundesilva) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/metrics/BallerinaMetricsObserver.java:40]: Can't we remove this now ?

File: observelib/observe/src/main/java/module-info.java [hunk 8]
```diff
@@ -3,4 +3,5 @@
     requires io.ballerina.runtime;
     requires opentracing.api;
     requires io.ballerina.config;
+    requires slf4j.api;
```
Comment (nadundesilva) [inline observelib/observe/src/main/java/module-info.java:6]: Do we need this ?

File: observelib/observe/src/main/java/org/ballerinalang/observe/nativeimpl/AddTagToMetrics.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+import java.util.HashMap;
+
+/**
+ * This function add tags to System Metrics.
+ * Custom tag is not included in the 'in progress-requests'
+ */
+public class AddTagToMetrics {
+
+    public static Object addTagToMetrics(Environment env, BString tagKey, BString tagValue) {
```
Comment (nadundesilva) [inline observelib/observe/src/main/java/org/ballerinalang/observe/nativeimpl/AddTagToMetrics.java:38]: We need to check if observability is enabled and immediately return `null` if it's disabled. (Similar to adding tracing tags)

File: observelib/observe/src/main/java/module-info.java (hunk 8 above)
Comment (LakshanKarunathilake) [reply observelib/observe/src/main/java/module-info.java:6]: This for logging the span not found error

**Review by nadundesilva — DISMISSED** [review]
LGTM

File: observelib/observe/src/main/java/org/ballerinalang/observe/nativeimpl/OpenTracerBallerinaWrapper.java [hunk 10]
```diff
@@ -149,29 +155,41 @@ public boolean finishSpan(Environment env, long spanId) {
...
-        ObserverContext observerContext = observerContextMap.get(spanId);
-        if (observerContext != null) {
-            observerContext.addTag(tagKey, tagValue);
-            return true;
+            span = (BSpan) observer.getProperty(TraceConstants.KEY_SPAN);
         } else {
-            return false;
+            ObserverContext observerContext = observerContextMap.get(spanId);
+            if (observerContext == null) {
+                log.info("Could not find the trace for given span id. Can not add tag {" + tagKey + ":" + tagValue +
```
Comment (manuranga) [inline observelib/observe/src/main/java/org/ballerinalang/observe/nativeimpl/OpenTracerBallerinaWrapper.java:182]: please extract this to a variable

**Review by manuranga — APPROVED** [review]

**Review by nadundesilva — APPROVED** [review]

**Pull request merged**